Add HomeKit Door accessory type

Dexwell opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR aims to make `cover` templates with the `device_class` set to `door` appear as actual doors in HomeKit. See #79363

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #79363
- This PR is related to issue: #79363
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26331

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Dexwell committed:
Add HomeKit Door accessory type

Dexwell commented:
I took a stab at implementing this myself—please bear with me as this is my first contribution to this project. I have not been able to test whether this achieves the desired result in HomeKit as I don't know how to test this in a development environment, but in a different project making these similar changes produced the correct result. Hopefully this is a good first start and someone else can get this PR across the finish line.

crazyfx1 commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -147,6 +147,11 @@ def get_accessory(  # noqa: C901
             a_type = "Window"
+        elif (
+            device_class == CoverDeviceClass.DOOR
+            and features & CoverEntityFeature.SET_POSITION
```
This is contrary to your implementation. The implementation only supports open / close (it even says so in the comment), but here you are checking if the entity only has set_position which is absolute positioning of the cover.

Dexwell replied:
Whoops, should be fixed in https://github.com/home-assistant/core/pull/80741/commits/ed6e7227d7927fb3b5d37d8637df3cfe037c3dc3

Dexwell committed:
Replace absolute positioning check with open/close

bdraco reviewed (dismissed):
All HomeKit accessories need 100% test coverage before we can merge them. Please add tests for this new accessory type.  Thanks!

Dexwell commented:
I've never written tests—can anyone else pick this up?

jredd commented:
@Dexwell If you based your code off of the garage integration you could just mimic the test coverage for the garage door. Take a look in the tests directory, it has a similar folder layout to the homeassistant directory.

jredd commented:
If I would be allowed to make changes to this I have the testing code made.

jredd committed:
feat: adding testing for the door type cover

bdraco commented:
@jredd Either @Dexwell can invite you to collaborate on his repo, or you might be better off opening a fresh PR

Dexwell commented:
@jredd That's awesome! Absolutely, I believe you can do so by creating a pull request to dexwell:homekit_door_accessory, correct?

jredd commented:
@Dexwell It took me a min to figure out how to do it with Github, I'm use to gitlab now and things are a bit weird for me. I have create a pr into your fork I believe let me know how it looks and if I need to create more changes.

Dexwell committed:
Merge pull request #1 from jredd/feat-tests-for-door-cover

feat: adding testing for the door type cover

jredd committed:
Merge branch 'dev' into homekit_door_accessory

bdraco commented on tests/components/homekit/test_type_covers.py:
```diff
@@ -129,6 +130,72 @@ async def test_garage_door_open_close(hass, hk_driver, events):
+    await hass.async_block_till_done()
+
+    assert acc.aid == 2
+    assert acc.category == 12  # GarageDoorOpener
```
Stale comment?

bdraco committed:
Merge branch 'dev' into homekit_door_accessory

Dexwell committed:
Fixed stale comment

bdraco committed:
lint

bdraco commented on homeassistant/components/homekit/type_covers.py:
```diff
@@ -188,6 +190,110 @@ def async_update_state(self, new_state):
+        state = self.hass.states.get(self.entity_id)
+
+        serv_door = self.add_preload_service(SERV_DOOR)
+        self.char_current_state = serv_door.configure_char(
+            CHAR_CURRENT_DOOR_STATE, value=0
+        )
+        self.char_target_state = serv_door.configure_char(
+            CHAR_TARGET_DOOR_STATE, value=0, setter_callback=self.set_state
+        )
```
Are these the right chars?
```
   "Door": {
      "OptionalCharacteristics": [
         "HoldPosition",
         "ObstructionDetected",
         "Name"
      ], 
      "RequiredCharacteristics": [
         "CurrentPosition",
         "PositionState",
         "TargetPosition"
      ], 
      "UUID": "00000081-0000-1000-8000-0026BB765291"
   },
```

Dexwell committed:
Base Door on Window instead of GarageDoorOpener

Given that WindowCovering, Window and Door accessory types in HomeKit have the same required characteristics

Dexwell commented:
I believe I initially did overthink the implementation. As @bdraco pointed out, `Door` has different characteristics than `GarageDoorOpener`. The new commit implements `Door` exactly like `Window` is implemented. This should work, given that `WindowCovering`, `Window` and `Door` accessory types all have the same required characteristics: `CurrentPosition`, `TargetPosition`, and `PositionState`.

@jredd Could you redo the testing code based on the window covering code instead of the garage door code?

Dexwell committed:
Base Door on Window

Dexwell committed:
Test Door the same way as Window

Dexwell commented:
Made an attempt at the testing code myself based on the Window code.

Dexwell committed:
Fixes for flake8

Dexwell committed:
Merge branch 'home-assistant:dev' into homekit_door_accessory

Dexwell committed:
Fix Black?

Dexwell committed:
Merge branch 'homekit_door_accessory' of https://github.com/Dexwell/core into homekit_door_accessory

Dexwell committed:
Merge branch 'home-assistant:dev' into homekit_door_accessory

Dexwell committed:
Merge latest changes

Dexwell committed:
Merge remote-tracking branch 'upstream/dev' into homekit_door_accessory

Dexwell committed:
Merge branch 'home-assistant:dev' into homekit_door_accessory

bdraco committed:
Merge branch 'dev' into homekit_door_accessory

bdraco commented:
Should be good to go once a docs PR is opened and linked to update https://www.home-assistant.io/integrations/homekit/#supported-components

Dexwell committed:
Merge branch 'dev' into homekit_door_accessory

Dexwell commented:
> Should be good to go once a docs PR is opened and linked to update https://www.home-assistant.io/integrations/homekit/#supported-components

Done!

bdraco committed:
Merge branch 'dev' into homekit_door_accessory

bdraco commented:
<img width="248" alt="Screenshot 2023-03-06 at 5 34 46 PM" src="https://user-images.githubusercontent.com/663432/223314421-9ac80162-10b9-40e4-b033-aceb72c73151.png">
<img width="512" alt="Screenshot 2023-03-06 at 5 33 24 PM" src="https://user-images.githubusercontent.com/663432/223314427-1a0e8514-6d46-42d6-b8b8-fae7aa52367c.png">

I changed one of my shades to display as a door and reloaded the bridge.

Its working as expected

bdraco commented:
Thanks @Dexwell

bdraco committed:
Merge branch 'dev' into homekit_door_accessory

bdraco commented:
We still need a test for the accessories file

```
 Check warning on line 155 in homeassistant/components/homekit/accessories.py

Codecov
/ codecov/patch
homeassistant/components/homekit/accessories.py#L155
Added line #L155 was not covered by tests
```

bdraco committed:
add missing coverage

bdraco commented:
I added the missing coverage.
